### PR TITLE
[MV-426] Porting & Benchmark Sigmoid Focal Loss kernel

### PR DESCRIPTION
**Pull request by BuiChiTrung**
- [x] Port 4 kernels to MiOpen: Fwd, Bwd, UnreducedFwd, UnreducedBwd.
- [x] Add unit tests for 4 kernels.
- [x] Add driver tests for 4 kernels.
- [x] Bench mark performance with ROCm: [link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B38FB5B9B-387C-4AD1-A594-8CC782CB3E90%7D&file=SigmoidFocalLoss.xlsx&action=default&mobileredirect=true)
- [x] Update docxygen documentation.

Reduce : 
|type    |Forward| Backward |
|--------|-------|-------|
|float32 | 2.61  | 4.27 |
|float16 | 2.18 | 3.82 | 
|bfloat16| 2.44  | 4.15 |

Unreduced: 
|type    | Forward | Backward |
|--------|-------|-------------|
|float32 | 5.09  | 3.39 |
|float16 | 4.74  | 3.03 |
|bfloat16| 5.27  | 3.40 |

### CODE REVIEW

**Comment by BuiChiTrung** [comment]
This PR has several changes compared to my last closed PR #20 for HingeEmbeddingLoss operation and other Loss kernel PRs. In previous PRs, we use 2 problem description structs and 2 invoke params structs for fwdReduced, fwdUnreduced kernels. However in this PR, I merged to 1 problem desc struct and 1 invoke param struct.

**Comment by BuiChiTrung** [comment]
Do we have to support computing the gradient for the target tensor in bwd kernels? I see some loss kernels in moDNN don't support whereas some support (BCELoss, ConsineEmbeddingLoss for instance).

**Comment by et16kr** [comment]
> Do we have to support computing the gradient for the target tensor in bwd kernels? I see some loss kernels in moDNN don't support whereas some support (BCELoss, ConsineEmbeddingLoss for instance).

I was delayed because I was looking through the history to investigate an issue I wasn't aware of.
When calculating target_grad in PyTorch's loss, we also added target_grad calculation to the modnn loss.
If a loss in PyTorch calculates target_grad, then it is necessary to include it in the modnn loss. 
Conversely, if a loss in PyTorch does not calculate target_grad, then it is not needed in modnn.
https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/FunctionsManual.cpp

It seems reasonable to follow the implementation in modnn. 
If you are unsure whether target_grad calculation is needed but missing, please check the corresponding PyTorch loss function.

File: driver/driver.hpp [hunk 1]
```diff
@@ -151,7 +151,8 @@ inline void PadBufferSize(size_t& sz, int datatype_sz)
            "pool[fp16], lrn[fp16], "
            "activ[fp16], softmax[fp16], bnorm[fp16], rnn[fp16], gemm[fp16], ctc, dropout[fp16], "
            "tensorop[fp16], reduce[fp16|fp64], layernorm[bfp16|fp16], sum[bfp16|fp16], "
-           "argmax[bfp16|fp16], groupnorm[bfp16|fp16], cat[bfp16|fp16]\n");
+           "argmax[bfp16|fp16], groupnorm[bfp16|fp16], cat[bfp16|fp16], "
+           "sigmoidfocalloss[bfp16|fp16|fp32] \n");
```
Comment (et16kr) [inline driver/driver.hpp:155]: fp32 seems to be a typo.
```bin/MIOpenDriver sigmoidfocallossfp32``` is not work.

**Comment by BuiChiTrung** [comment]
> > Do we have to support computing the gradient for the target tensor in bwd kernels? I see some loss kernels in moDNN don't support whereas some support (BCELoss, ConsineEmbeddingLoss for instance).
> 
> I was delayed because I was looking through the history to investigate an issue I wasn't aware of. When calculating target_grad in PyTorch's loss, we also added target_grad calculation to the modnn loss. If a loss in PyTorch calculates target_grad, then it is necessary to include it in the modnn loss. Conversely, if a loss in PyTorch does not calculate target_grad, then it is not needed in modnn. https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/FunctionsManual.cpp
> 
> It seems reasonable to follow the implementation in modnn. If you are unsure whether target_grad calculation is needed but missing, please check the corresponding PyTorch loss function.

I figured out that only loss kernel with int/long target tensor, the dtarget is not computed. I have added the code to compute target tensor gradient.

File: include/miopen/miopen.h [hunk 2]
```diff
@@ -512,7 +512,7 @@ typedef enum
     miopenActivationABS      = 5, /*!< Absolute value \f$abs(x)\f$ */
     miopenActivationPOWER = 6, /*!< Scaled and shifted power \f$(\alpha + \beta * x)^{gamma}\f$ */
     miopenActivationCLIPPEDRELU =
-        7, /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
+        7,                     /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
```
Comment (et16kr) [inline include/miopen/miopen.h]: please check clang-format

File: test/gtest/sigmoid_focal_loss.cpp [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+};
+
+struct SigmoidFocalLossUnreducedBackwardTestBFloat16 : SigmoidFocalLossUnreducedBwdTest<bfloat16>
+{
+};
+}; // namespace sigmoidfocalloss
+
+using namespace sigmoidfocalloss;
+
+#ifdef TEST_FWD_REDUCED
```
Comment (et16kr) [inline test/gtest/sigmoid_focal_loss.cpp:103]: There is define TEST_FWD_REDUCED right above it. So it seems meaningless.
Please explain why you did this

File: test/gtest/sigmoid_focal_loss.cpp [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+TEST_P(SigmoidFocalLossForwardTestFloat32, SigmoidFocalLossForwardTest)
+{
+    if(miopen::IsEnabled(ENV(MIOPEN_TEST_ALL)))
```
Comment (et16kr) [inline test/gtest/sigmoid_focal_loss.cpp]: We have recently received a request from AMD.
Please refer to the following
https://github.com/ROCm/MIOpen/issues/1843
https://github.com/ROCm/MIOpen/pull/2883#discussion_r1616401109

**Comment by et16kr** [comment]
Have you measured performance after adding Target grad?
I think it would be better to optionally select target grad in driver.(default: off)
This is because grad is not generally used in target tensor,
and performance measurements may vary if target grad is included. 
(I think gtest is fine as is.)

I think you made the right choice to include target grad.
But I'll ask others as well.(Their response is OK!!)

File: src/include/miopen/tensor_view_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            tensor_view.stride[i] = (i == 0 ? 1 : strides[i - 1]);
+            tensor_view.size[i]   = 1;
+        }
+    }
+    return tensor_view;
+}
+
+template <int N>
+inline void slice_tv(tensor_view_t<N>& tensor_view, int32_t sliceCount, const int32_t* slices)
```
Comment (et16kr) [inline src/include/miopen/tensor_view_utils.hpp:64]: slice_tv is not used in this PR.
If it was used in another PR, could you please let me know so I can refer to it?

File: driver/sigmoid_focal_loss_driver.hpp [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+    }
+
+    return miopenStatusSuccess;
+}
+
+template <typename TIO>
+int SigmoidFocalLossDriver<TIO>::VerifyBackward()
+{
+    RunBackwardCPU();
+#ifdef DEBUGGING
```
Comment (et16kr) [inline driver/sigmoid_focal_loss_driver.hpp]: Please delete the debugging output.
Please let me know if you still need it, and please delete it when uploading it to AMD.

File: src/kernels/MIOpenSigmoidFocalLoss.cpp [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+    {
+        FLOAT_ACCUM alpha_t = alpha * t + (1 - alpha) * (1 - t);
+        loss                = alpha_t * loss;
+    }
+
+    workspace[gid] = CVT_ACCUM2FLOAT(loss / divisor);
+}
+
+extern "C" __global__ void SigmoidFocalLossFwd(const IN_OUT_TYPE* input,
+                                               IN_OUT_TYPE* target,
```
Comment (et16kr) [inline src/kernels/MIOpenSigmoidFocalLoss.cpp:81]: Tensors are of the same dtype.
A dtype check of the tensor is required.
Please let me know if I can't find it. (I only found checkSameLength.)

File: driver/sigmoid_focal_loss_driver.hpp (hunk 6 above)
Comment (BuiChiTrung) [reply driver/sigmoid_focal_loss_driver.hpp]: It's is removed

File: src/include/miopen/tensor_view_utils.hpp (hunk 5 above)
Comment (BuiChiTrung) [reply src/include/miopen/tensor_view_utils.hpp:64]: It's used in Seungman's [PR](https://github.com/ROCm/MIOpen/pull/2883/files#diff-ccc972090dd3e61487d0ad364fb34d4aaf8f7c6911c878988654fe7173f3058f)

File: test/gtest/sigmoid_focal_loss.cpp (hunk 3 above)
Comment (BuiChiTrung) [reply test/gtest/sigmoid_focal_loss.cpp:103]: While benchmarking, testing, and debugging we may want to run only a specific set of tests instead of all tests. I added these macros to make it easier to enable/disable all tests of a kernel.

File: include/miopen/miopen.h (hunk 2 above)
Comment (BuiChiTrung) [reply include/miopen/miopen.h]: Fixed it.

File: driver/driver.hpp (hunk 1 above)
Comment (BuiChiTrung) [reply driver/driver.hpp:155]: Yes, it is a typo. Driver with float32 is invoked like this: `bin/MIOpenDriver sigmoidfocalloss`

**Comment by BuiChiTrung** [comment]
> Have you measured performance after adding Target grad? I think it would be better to optionally select target grad in driver.(default: off) This is because grad is not generally used in target tensor, and performance measurements may vary if target grad is included. (I think gtest is fine as is.)
> 
> I think you made the right choice to include target grad. But I'll ask others as well.

The benchmark result for target grad is updated in the PR desc. Also, i have added a flag to disable/enable target grad in driver, this required changing the tensor_view_utils.hpp as follow:
```
for(size_t i = 0; i < N; ++i)
    {
        if(dims.empty())
        {
            tensor_view.stride[i] = 0;
            tensor_view.size[i]   = 0;
        }
        ...
    }
```

File: src/include/miopen/tensor_view_utils.hpp (hunk 5 above)
Comment (et16kr) [reply src/include/miopen/tensor_view_utils.hpp:64]: OK I see it.

File: src/kernels/MIOpenSigmoidFocalLoss.cpp (hunk 7 above)
Comment (BuiChiTrung) [reply src/kernels/MIOpenSigmoidFocalLoss.cpp:81]: It's my bad. I have added the check code in problem description constructor

File: test/gtest/sigmoid_focal_loss.cpp (hunk 4 above)
Comment (BuiChiTrung) [reply test/gtest/sigmoid_focal_loss.cpp]: I have updated the test invoke condition

File: driver/sigmoid_focal_loss_driver.hpp [hunk 8]
```diff
@@ -0,0 +1,774 @@
...
+
+template <typename TIO>
+int SigmoidFocalLossDriver<TIO>::GetandSetData()
+{
+    auto inDims              = inflags.GetValueTensor("dim-lengths").lengths;
+    alpha        = inflags.GetValueDouble("alpha");
+    gamma        = inflags.GetValueDouble("gamma");
+    isContiguous = inflags.GetValueInt("is-contiguous") == 1 ? true : false;
+    isTargetGradientComputed = inflags.GetValueInt("target-gradient") == 1 ? true : false;
+    reduction    = static_cast<miopenLossReductionMode_t>(inflags.GetValueInt("reduction"));
```
Comment (et16kr) [inline driver/sigmoid_focal_loss_driver.hpp]: clang-format is required.
If you use a github hook, it will be checked every time you commit.
```
./.githooks/install
git config core.hooksPath .githooks
```

Comment (BuiChiTrung) [reply driver/sigmoid_focal_loss_driver.hpp]: I migrated code to a new node yesterday and forgot to setup the githooks and clang-format in IDE. I have updated it.

File: test/gtest/sigmoid_focal_loss.hpp [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+};
+
+inline std::vector<SigmoidFocalLossTestCase> SigmoidFocalLossTestConfigs()
+{
+    return {
+        SigmoidFocalLossTestCase({4000}),            // 1D cont
+        SigmoidFocalLossTestCase({100, 500}),        // 2D cont
+        SigmoidFocalLossTestCase({100, 500}, false), // 2D non-cont
+        // SigmoidFocalLossTestCase({2, 2}), // 2D non-cont
+        // SigmoidFocalLossTestCase({2, 2}, false), // 2D non-cont
```
Comment (et16kr) [inline test/gtest/sigmoid_focal_loss.hpp]: If you do not use it, please remove it.

File: include/miopen/miopen.h [hunk 10]
```diff
@@ -4950,6 +4950,98 @@ MIOPEN_EXPORT miopenStatus_t miopenCTCLoss(miopenHandle_t handle,
...
+ */
+MIOPEN_EXPORT miopenStatus_t
+miopenGetSigmoidFocalLossForwardWorkspaceSize(miopenHandle_t handle,
+                                              miopenTensorDescriptor_t inputDesc,
+                                              miopenTensorDescriptor_t targetDesc,
+                                              miopenTensorDescriptor_t outputDesc,
+                                              miopenLossReductionMode_t reduction,
+                                              size_t* sizeInBytes);
+
+/*! @brief Execute a SigmoidFocalLoss reduced forward layer
```
Comment (et16kr) [inline include/miopen/miopen.h:4982]: Since this is an API that includes both reduced and unreduced, 
I think 'reduced' should be removed from the function comment.

Comment (et16kr) [inline include/miopen/miopen.h]: This is an API that includes both reduced and unreduced.
I think "reduced" should be removed in function comment.

File: test/gtest/sigmoid_focal_loss.hpp [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+        dOutput             = tensor<TIO>{in_dims, in_strides}.generate(dOut_gen_value);
+
+        dInput = tensor<TIO>{in_dims};
+        std::fill(dInput.begin(), dInput.end(), 0);
+
+        dInputHost = tensor<TIO>{in_dims};
+        std::fill(dInputHost.begin(), dInputHost.end(), 0);
+
+        dTarget = tensor<TIO>{in_dims};
+        std::fill(dInput.begin(), dInput.end(), 0);
```
Comment (et16kr) [inline test/gtest/sigmoid_focal_loss.hpp]: maybe ```std::fill(dTarget.begin(), dTarget.end(), 0);```

File: test/gtest/sigmoid_focal_loss.hpp [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+        dOutput[0] = prng::gen_descreet_uniform_sign<TIO>(0.1, 50);
+
+        dInput = tensor<TIO>{in_dims};
+        std::fill(dInput.begin(), dInput.end(), 0);
+
+        dInputHost = tensor<TIO>{in_dims};
+        std::fill(dInputHost.begin(), dInputHost.end(), 0);
+
+        dTarget = tensor<TIO>{in_dims};
+        std::fill(dInput.begin(), dInput.end(), 0);
```
Comment (et16kr) [inline test/gtest/sigmoid_focal_loss.hpp]: ```std::fill(dTarget.begin(), dTarget.end(), 0);```

Comment (BuiChiTrung) [reply test/gtest/sigmoid_focal_loss.hpp]: You're right. Updated

File: test/gtest/sigmoid_focal_loss.hpp (hunk 11 above)
Comment (BuiChiTrung) [reply test/gtest/sigmoid_focal_loss.hpp]: You're right. Updated

File: include/miopen/miopen.h (hunk 10 above)
Comment (BuiChiTrung) [reply include/miopen/miopen.h]: It is removed.

File: test/gtest/sigmoid_focal_loss.hpp (hunk 9 above)
Comment (BuiChiTrung) [reply test/gtest/sigmoid_focal_loss.hpp]: It is removed.

File: include/miopen/miopen.h (hunk 2 above)
Comment (BuiChiTrung) [inline include/miopen/miopen.h:515]: My IDE keeps adding this empty space, as my VPN suddenly stopped working, I will fix it tomorrow.

**Comment by et16kr** [comment]
Thank you for your hard work.
